File: app/src/main/java/com/thing/collab/LoginActivity.java
package com.thing.collab;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.*;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.thing.collab.handler.UserHandler;

import static com.thing.collab.handler.UserHandler.FAILED;
import static com.thing.collab.handler.UserHandler.UPDATED;

public class LoginActivity extends AppActivity {

    private Context context = this;

    private EditText inputEmail, inputPassword;
    private FirebaseAuth auth;
    private ProgressBar progressBar;
    private Button btnLogin;

    static final String TAG = "LoginActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        auth = FirebaseAuth.getInstance();

        if (auth.getCurrentUser() != null) {
            UserHandler.getInstance(this.getApplicationContext());
            startActivity(new Intent(LoginActivity.this, MainActivity.class));
        }

        setContentView(R.layout.activity_login);
        inputEmail = findViewById(R.id.email);
        inputPassword = findViewById(R.id.password);
        progressBar = findViewById(R.id.progressBar);
        btnLogin = findViewById(R.id.btn_login);


        auth = FirebaseAuth.getInstance();
        btnLogin.setOnClickListener(v -> {
            String email = inputEmail.getText().toString();
            final String password = inputPassword.getText().toString();

            if (TextUtils.isEmpty(email)) {
                Toast.makeText(getApplicationContext(), "Enter email address!", Toast.LENGTH_SHORT).show();
                return;
            }

            if (TextUtils.isEmpty(password)) {
                Toast.makeText(getApplicationContext(), "Enter password!", Toast.LENGTH_SHORT).show();
                return;
            }

            progressBar.setVisibility(View.VISIBLE);

            //authenticate user
            auth.signInWithEmailAndPassword(email, password)
                    .addOnCompleteListener(LoginActivity.this, new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull Task<AuthResult> task) {
                            // If sign in fails, display a message to the user. If sign in succeeds
                            // the auth state listener will be notified and logic to handle the
                            // signed in user can be handled in the listener.
                            progressBar.setVisibility(View.GONE);
                            if (!task.isSuccessful()) {
                                // there was an error
                                if (password.length() < 6) {
                                    inputPassword.setError("Enter Password with minimum 6 characters");
                                } else {
                                    Log.e(TAG, task.getException().getMessage());
                                    Toast.makeText(LoginActivity.this, "Authentication Failed", Toast.LENGTH_LONG).show();
                                }
                            } else {
                                UserHandler.getInstance(LoginActivity.this.getApplicationContext()).updateUserFromAuth(auth.getUid(), (user, flag) -> {
                                    switch (flag) {
                                        case UPDATED:
                                            Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                                            startActivity(intent);
                                            break;
                                        case FAILED:
                                            Toast.makeText(LoginActivity.this, "Authentication Failed", Toast.LENGTH_LONG).show();
                                            break;
                                    }
                                });

                            }
                        }
                    });
        });

    }
}

File: app/src/main/java/com/thing/allied/LoginActivity.kt
package com.thing.allied

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*

import com.google.firebase.auth.FirebaseAuth
import com.thing.allied.handler.UserHandler

import kotlinx.android.synthetic.main.activity_login.*

class LoginActivity : AppActivity() {

    private var auth: FirebaseAuth? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        auth = FirebaseAuth.getInstance()
        loginBtn.setOnClickListener {

            if (loginInputEmail.text.isNullOrEmpty()) {
                Toast.makeText(applicationContext, "Enter email address!", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            if (loginInputPass.text.isNullOrEmpty()) {
                Toast.makeText(applicationContext, "Enter password!", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            progressBar!!.visibility = View.VISIBLE

            //authenticate user
            auth!!.signInWithEmailAndPassword(loginInputEmail.text.toString(), loginInputPass.text.toString())
                .addOnCompleteListener(this@LoginActivity) { task ->
                    // If sign in fails, display a message to the user. If sign in succeeds
                    // the auth state listener will be notified and logic to handle the
                    // signed in user can be handled in the listener.
                    progressBar!!.visibility = View.GONE
                    if (!task.isSuccessful) {
                        // there was an error
                        if (loginInputPass.text.length < 6) {
                            loginInputPass!!.error = "Enter Password with minimum 6 characters"
                        } else {
                            Log.e(TAG, task.exception!!.message)
                            Toast.makeText(this@LoginActivity, "Authentication Failed", Toast.LENGTH_LONG).show()
                        }
                    } else {
                        UserHandler.updateUserFromAuth(auth!!.uid!!, onUserUpdate = { user ->
                            if (user == null) {
                                Toast.makeText(this, "User not found", Toast.LENGTH_SHORT).show()
                            } else {
                                finish()
                            }
                        })
                    }
                }
        }

    }

    companion object {
        private const val TAG = "LoginActivity"
    }
}
